File: algo/week02/1162-as-far-from-land-as-possible/solution.ts
/**
 * 1162. 地图分析
 * https://leetcode.cn/problems/as-far-from-land-as-possible/
 *
 * 多源 BFS：所有陆地同时入队，逐层扩展到海洋，最后一层距离即答案。
 */

export function maxDistance(grid: number[][]): number {
  const n = grid.length;
  const queue: [number, number][] = [];

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i]![j] === 1) queue.push([i, j]);
    }
  }

  if (queue.length === 0 || queue.length === n * n) return -1;

  const dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]];
  let dist = -1;
  let head = 0;

  while (head < queue.length) {
    const size = queue.length - head;
    for (let s = 0; s < size; s++) {
      const [r, c] = queue[head++]!;
      for (const [dr, dc] of dirs) {
        const nr = r + dr!;
        const nc = c + dc!;
        if (nr >= 0 && nr < n && nc >= 0 && nc < n && grid[nr]![nc] === 0) {
          grid[nr]![nc] = 1;
          queue.push([nr, nc]);
        }
      }
    }
    dist++;
  }

  return dist;
}

File: algo/week02/1162-as-far-from-land-as-possible/solution.rs
//! 1162. 地图分析 — 多源 BFS
//! https://leetcode.cn/problems/as-far-from-land-as-possible/

use std::collections::VecDeque;

struct Solution;

impl Solution {
    pub fn max_distance(mut grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let mut queue: VecDeque<(usize, usize)> = VecDeque::new();

        for i in 0..n {
            for j in 0..n {
                if grid[i][j] == 1 {
                    queue.push_back((i, j));
                }
            }
        }

        if queue.is_empty() || queue.len() == n * n {
            return -1;
        }

        let dirs: [(i32, i32); 4] = [(0, 1), (0, -1), (1, 0), (-1, 0)];
        let mut dist = -1;

        while !queue.is_empty() {
            let size = queue.len();
            for _ in 0..size {
                let (r, c) = queue.pop_front().unwrap();
                for &(dr, dc) in &dirs {
                    let nr = r as i32 + dr;
                    let nc = c as i32 + dc;
                    if nr >= 0 && nr < n as i32 && nc >= 0 && nc < n as i32 {
                        let (nru, ncu) = (nr as usize, nc as usize);
                        if grid[nru][ncu] == 0 {
                            grid[nru][ncu] = 1;
                            queue.push_back((nru, ncu));
                        }
                    }
                }
            }
            dist += 1;
        }

        dist
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn max_distance_matches_shared_cases() {
        assert_eq!(
            Solution::max_distance(vec![vec![1, 0, 1], vec![0, 0, 0], vec![1, 0, 1]]),
            2
        );
        assert_eq!(
            Solution::max_distance(vec![vec![1, 0, 0], vec![0, 0, 0], vec![0, 0, 0]]),
            4
        );
        assert_eq!(
            Solution::max_distance(vec![vec![1, 1, 1], vec![1, 1, 1], vec![1, 1, 1]]),
            -1
        );
        assert_eq!(
            Solution::max_distance(vec![vec![0, 0, 0], vec![0, 0, 0], vec![0, 0, 0]]),
            -1
        );
    }
}

fn main() {}
